typed-lambda-calculus/article.md

```diff
@@ -0,0 +1,996 @@
+\end{array}
+$$
+
+The $\mathbf{over}$ expression declares an overloaded identifier.
+In the scope of this declaration, zero or more $\mathbf{inst}$ declarations are allowed.
+
+Both $\mathbf{over}$ and $\mathbf{inst}$ expressions must be explicitly typed.
+
+The type of an overloaded identifier in $\mathbf{inst}$ must be an _instance_ of the type in the matching $\mathbf{over}$ declaration.
+_Instance_ in this context is essentially the same $\sqsubseteq$ relation defined in HM, but extended for the predicated types.
+
+If you're familiar with type classes in Haskell, this syntax might seem slightly alien (probably more so if you're not), but the translation is rather straightforward.
+
+In Haskell, we would write
+
+```haskell
+member :: Eq a => [a] -> a -> Bool
+```
+
+In this extended lambda-calculus we would have instead
+
+$$
+member : \forall \alpha.\;(eq : Eq\;\alpha).\;[\alpha] \to \alpha \to Bool
+$$
+
+The paper also provides typing and translation rules, but those are somewhat involved, so we'll gloss over them here.
+The paper itself is very readable, so I encourage you to read it if interested.
+
+I will provide an example from the paper, however:
+
+$$
+\begin{array}{l}
+\mathbf{over}\;numD : \forall\alpha.\;Num\;\alpha\;\mathbf{in}\\
+\mathbf{inst}\;numD : Num\;Int = (addInt, mulInt, negInt)\;\mathbf{in}\\
+\mathbf{inst}\;numD : Num\;Float = (addFloat, mulFloat, negFloat)\;\mathbf{in}\\
+\mathbf{let}\;(+) = fst\;numD\;\mathbf{in}\\
+\mathbf{let}\;(*) = snd\;numD\;\mathbf{in}\\
+\mathbf{let}\;negate = thd\;numD\;\mathbf{in}\\
+\mathbf{let}\;square = \lambda x.\;x * x\;\mathbf{in}\\
+square\;3
+\end{array}
+$$
+
+Notice how the dictionary (in this example for `Num`, containing only three methods and represented as a tuple) ends up being used explicitly.
+
+There is however an issue with this calculus, namely, used as described here, a principal type might not exist.
+
+For example, consider
+
+$$
+\begin{array}{l}
+\mathbf{over}\;eq : \forall\alpha.\;Eq\;\alpha\;\mathbf{in}\\
+\mathbf{inst}\;eq : Eq\;Int = eqInt\;\mathbf{in}\\
+\mathbf{inst}\;eq : Eq\;Char = eqChar\;\mathbf{in}\\
+eq
+\end{array}
+$$
+
+Now what is the type of this expression as a whole?
+Note we can't, practically speaking, reference out-of-scope predicated types to type it.
+Hence we need to make a decision whether it's type is the type of $eqInt$ or $eqChar.$
+But there is no principal type here, both choices are equal, neither type is smaller than the other.
+
+<p><details>
+<summary>On typing type classes</summary>
+
+In the paragraph above, I'm glossing over a lot of details for simplicity.
+In particular, I'm conflating typing and translation to HM, which are two distinct steps.
+Still, this gets the core idea accross well enough.
+
+***
+
+</details></p>
+
+To work around this issue, it is required that $\mathbf{over}$ and $\mathbf{inst}$ must have global scope.
+That's why in Haskell we can't have local instance declarations.
```

Comment:
Given that there is an association between GHC and Haskell (so it feels a bit confusing), and the details feel wordy here, I'd propose something like this perhaps: 

> in X, we can have local instance declarations only through a lot of dark and arcane compiler magic.

Reply:
Uh... Haskell is a standard, GHC is one implementation (with a lot of extensions).

But really, there's no way to get local instances even with GHC extensions, getting those needs some `unsafeCoerce` finagling using knowledge of GHC's runtime representations. So it's "kind of" possible with a lot of magic™.